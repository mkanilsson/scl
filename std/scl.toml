[package]
name = "std"
description = "Scl's standard library"
type = "lib"
