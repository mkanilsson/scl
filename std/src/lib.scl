extern proc exit(i32) !
