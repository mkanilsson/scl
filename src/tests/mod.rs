use std::{env, fs, path::PathBuf, process::Command};

use insta::glob;
use miette::{GraphicalReportHandler, GraphicalTheme};
use nanoid::nanoid;

use crate::{codegen::Codegen, package::Package, typechecker::Checker};

#[test]
fn runtime() {
    let std_package = Package::from_path("std", "std".into()).unwrap();
    let std_package = std_package.parse().unwrap();

    fn generate_tempfile(extension: &str) -> PathBuf {
        let name = format!("{}.{extension}", nanoid!());
        env::temp_dir().join(name)
    }

    glob!("sources/runtime/*", |file| {
        let main_package = Package::from_file(file.into()).unwrap();
        let main_package = main_package.parse().unwrap();

        let mut checker = Checker::new();
        let checked_std_package = checker.add_package(&std_package, &vec![]).unwrap();

        let checked_main_package = checker
            .add_package(&main_package, &vec![(
                "std".into(),
                checked_std_package.package_id,
            )])
            .unwrap();

        insta::assert_debug_snapshot!(checked_main_package);

        let all_units = checked_std_package
            .units
            .into_iter()
            .chain(checked_main_package.units.into_iter())
            .collect::<Vec<_>>();

        let mut codegener = Codegen::new(all_units, checker);
        let code = codegener.generate();

        insta::assert_snapshot!(code);

        let tmp = generate_tempfile("ssa");
        let asm = generate_tempfile("S");
        let executable = generate_tempfile("out");

        std::fs::write(&tmp, code).unwrap();

        let qbe_cmd = Command::new("qbe")
            .arg("-o")
            .arg(&asm)
            .arg(tmp)
            .output()
            .unwrap();

        if !qbe_cmd.status.success() {
            let stdout = String::from_utf8_lossy(&qbe_cmd.stdout);
            let stderr = String::from_utf8_lossy(&qbe_cmd.stderr);

            panic!("QBE failed:\nOut:\n{stdout}Err:\n{stderr}");
        }

        let gcc_cmd = Command::new("gcc")
            .arg("-o")
            .arg(&executable)
            .arg(&asm)
            .output()
            .unwrap();

        if !gcc_cmd.status.success() {
            let stdout = String::from_utf8_lossy(&gcc_cmd.stdout);
            let stderr = String::from_utf8_lossy(&gcc_cmd.stderr);

            panic!("gcc failed:\nOut:\n{stdout}Err:\n{stderr}");
        }

        let run_cmd = Command::new(executable).output().unwrap();

        let stdout = String::from_utf8_lossy(&run_cmd.stdout);
        let stderr = String::from_utf8_lossy(&run_cmd.stderr);

        insta::assert_snapshot!(stdout);
        insta::assert_snapshot!(stderr);
        insta::assert_snapshot!(run_cmd.status);
    });
}

#[test]
fn typecheck_errors() {
    let std_package = Package::from_path("std", "std".into()).unwrap();
    let std_package = std_package.parse().unwrap();

    glob!("sources/typecheck_error/*", |file| {
        let main_package = Package::from_file(file.into()).unwrap();
        let main_package = main_package.parse().unwrap();

        let mut checker = Checker::new();
        let checked_std_package = checker.add_package(&std_package, &vec![]).unwrap();

        let err = checker
            .add_package(&main_package, &vec![(
                "std".into(),
                checked_std_package.package_id,
            )])
            .err()
            .unwrap();

        let mut buf = String::new();
        GraphicalReportHandler::new_themed(GraphicalTheme::none())
            .render_report(&mut buf, miette::Report::new(err).as_ref())
            .unwrap();

        insta::assert_snapshot!(buf);
    });
}

#[test]
fn parse_expr() {
    glob!("sources/parse_expr/*", |file| {
        let source = fs::read_to_string(file).unwrap();
        let expr = crate::new_parser::Parser::parse_expr(&source);

        insta::assert_debug_snapshot!(expr);
    });
}

#[test]
fn parse_stmt() {
    glob!("sources/parse_stmt/*", |file| {
        let source = fs::read_to_string(file).unwrap();
        let expr = crate::new_parser::Parser::parse_stmt(&source);

        insta::assert_debug_snapshot!(expr);
    });
}
