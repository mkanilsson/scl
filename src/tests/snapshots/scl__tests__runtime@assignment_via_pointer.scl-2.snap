---
source: src/tests/mod.rs
expression: code
input_file: src/tests/sources/runtime/assignment_via_pointer.scl
---
type :main.Thing = { w }
export function w $std.math.add(w %a, w %b) {
@start
	%.ss.0 =l alloc4 4
	%.ss.1 =l alloc4 4
	storew %a, %.ss.0
	storew %b, %.ss.1
@body
	%.sv.1 =w loadw %.ss.0
	%.sv.2 =w loadw %.ss.1
	%.binop.3 =w add %.sv.1, %.sv.2
	ret %.binop.3
}
export function w $main() {
@start
	%.ss.0 =l alloc4 4
	%.ss.1 =l alloc8 8
	%.ss.2 =l alloc8 8
	%.ss.3 =l alloc4 4
	%.ss.4 =l alloc4 4
	%.ss.5 =l alloc4 4
	%.ss.6 =l alloc8 8
	%.ss.7 =l alloc8 8
	%.ss.8 =l alloc4 4
	%.ss.9 =l alloc4 4
@body
	storew 10, %.ss.0
	storel %.ss.0, %.ss.1
	%.sv.4 =l loadl %.ss.1
	storew 20, %.sv.4
	%.sv.7 =w loadw %.ss.0
	%.printf.return_value.5 =w call $printf(l $.str.6, ..., w %.sv.7)
	# Store value into Thing.value (offset: 0)
	%.offset.8 =l add %.ss.4, 0
	storew 30, %.offset.8
	blit %.ss.4, %.ss.5, 4
	storel %.ss.5, %.ss.6
	%.sv.9 =l loadl %.ss.6
	# Store value into Thing.value (offset: 0)
	%.offset.10 =l add %.ss.8, 0
	storew 40, %.offset.10
	blit %.ss.8, %.sv.9, 4
	%.offset.13 =l add %.ss.5, 0
	%.member_access.14 =w loadw %.offset.13
	%.printf.return_value.11 =w call $printf(l $.str.12, ..., w %.member_access.14)
	ret 0
}
data $.str.6 = { b "a = %d\n", b 0 }
data $.str.12 = { b "c.value = %d\n", b 0 }
